net: DHCPv4 client: Add support for encapsulated vendor specific options

maass-hamburg opened the pull request:
Similarly to the already existing dhcpv4 option callbacks, I implemented callbacks for encapsulated vendor-specific options in DHCP option 43 as described by RFC 2132.  It allows the direct use of the encapsulated vendor-specific options instead of having to implement a parser of the encapsulated vendor-specific options in the application code.

A test is also provided.

maass-hamburg committed:
net: dhcpv4: add vendor specific option callback

In certain scenarios, it may be necessary to get values of additional
options from the application layer. With this patch, this can be
accomplished by registering a callback with the DHCP client.

This change has been tested using the posix build in qemu.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

jukkar commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -810,6 +817,68 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
+
+			if (length >= 3) {
+
+				struct net_dhcpv4_option_callback *vs_cb, *vs_tmp;
```
As there are multiple level of `{}` blocks below, I would suggest to make the added code as a function that we could call here. This would help the readability of the code.

jukkar commented on include/zephyr/net/dhcpv4.h:
```diff
@@ -163,6 +169,48 @@ int net_dhcpv4_remove_option_callback(struct net_dhcpv4_option_callback *cb);
+				       net_dhcpv4_option_callback_handler_t handler, uint8_t option,
+				       void *data, size_t max_length)
+{
+	__ASSERT(callback, "Callback pointer should not be NULL");
+	__ASSERT(handler, "Callback handler pointer should not be NULL");
+	__ASSERT(data, "Data pointer should not be NULL");
+
+	callback->handler = handler;
+	callback->option = option;
+	callback->data = data;
+	callback->max_length = max_length;
```
As this is application interfacing API, I would suggest that we check the parameters properly and return error codes etc.

m-byte replied:
@jukkar Since I had asked Fin to base this code on my option callback code (#57918), this part, as well as the one mentioned in your slist comment are pretty much exact copies of the code for regular callbacks.
Since your comments also apply to that already mearged code, should we redo those as well? If so, should that be a separate PR or just part of this one?

jukkar replied:
Ah, I missed that the existing function was doing the same thing. Hmm, lets leave this like this for now and deal with this later then.

m-byte replied:
Probably for the better, since my code in turn was based on the gpio callback API (https://github.com/zephyrproject-rtos/zephyr/blob/40b53d5d774705d27183aaa1e42f2f322250da10/include/zephyr/drivers/gpio.h#L1692)..

jukkar commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -48,6 +48,10 @@ static struct net_mgmt_event_callback mgmt4_cb;
 #endif
 
+#if defined(CONFIG_NET_DHCPV4_OPTION_CALLBACKS_VENDOR_SPECIFIC)
+static sys_slist_t option_vendor_callbacks;
```
There is a macro that can be used to init the slist, we should use it here.
```
static sys_slist_t option_vendor_callbacks = SYS_SLIST_STATIC_INIT(&option_vendor_callbacks);
```

rlubos commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -810,6 +867,19 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
+			if (length >= 3) {
+				net_pkt_cursor_backup(pkt, &backup);
+				dhcpv4_parse_option_vendor(pkt, iface, msg_type, length);
+				net_pkt_cursor_restore(pkt, &backup);
```
Why do we need to backup/restore cursor here? IMO `dhcpv4_parse_option_vendor()` should advance the cursor, as it parses consecutive option data. If it encounters problem (malformed option, for example no END marker), it should return an error.

rlubos commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -667,6 +671,56 @@ static void dhcpv4_timeout(struct k_work *work)
+			NET_DBG("Vendor-specific options_end");
+			return 0;
+		}
+		len_counter--;
```
Can't we just decrement `length` directly? I see no other use of that variable.

rlubos commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -667,6 +671,56 @@ static void dhcpv4_timeout(struct k_work *work)
+
+	while (!net_pkt_read_u8(pkt, &vs_type) && len_counter > 0) {
+		if (vs_type == DHCPV4_OPTIONS_END) {
+			NET_DBG("Vendor-specific options_end");
```
Shouldn't we check if `len_counter` (`length`) is 0 here? Otherwise, it means END marker was read too early (malformed option?)

rlubos commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -667,6 +671,56 @@ static void dhcpv4_timeout(struct k_work *work)
+		net_pkt_skip(pkt, len);
+		len_counter = len_counter - len;
+	}
+	return 0;
```
Same here, if we exit here, it means we did not read END marker, IMO that's an error condition.

maass-hamburg replied:
According to the RFC 2132 the existence of the end maker in encapsulated vendor options is not mandatory 
`3) Code 255 (END), if present, signifies the end of the
         encapsulated vendor extensions, not the end of the vendor
         extensions field. If no code 255 is present, then the end of
         the enclosing vendor-specific information field is taken as the
         end of the encapsulated vendor-specific extensions field.`

rlubos replied:
Thanks for clarifying, I got confused as END marker is mandatory in the DHCP option list, but apparently that's no the case for the option sub-list.

rlubos commented on include/zephyr/net/dhcpv4.h:
```diff
@@ -64,7 +64,8 @@ enum net_dhcpv4_msg_type {
 };
 
-#ifdef CONFIG_NET_DHCPV4_OPTION_CALLBACKS
+#if defined(CONFIG_NET_DHCPV4_OPTION_CALLBACKS) ||                                                 \
```
I just wonder if it'd make sense to make `CONFIG_NET_DHCPV4_OPTION_CALLBACKS_VENDOR_SPECIFIC` dependent on `CONFIG_NET_DHCPV4_OPTION_CALLBACKS` (so that you need to enable option callbacks to get vendor-specific extensions). That should reduce ifdef slicing of the code a bit, that long `if defined(this) || defined(that)` condition looks a bit worrying.

m-byte replied:
Well, on the one hand, the functionality does not really depend on option callbacks being enabled. On the other hand I agree that from a user standpoint encapsulated option callbacks are just a specific type of option callbacks, even if their parsing is done separately.
Overall, I am also in favor of making encapsulated vendor options a suboption in Kconfig.

jukkar commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -667,6 +671,71 @@ static void dhcpv4_timeout(struct k_work *work)
+	uint8_t len;
+	uint8_t type;
+
+	if (length >= 3) {
```
The debug msg says too short so shouldn't this be `<`?

maass-hamburg replied:
yes

jukkar commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -667,6 +671,71 @@ static void dhcpv4_timeout(struct k_work *work)
+
+	if (length >= 3) {
+		NET_ERR("Vendor-specific option parsing, length too short");
+		net_pkt_skip(pkt, length);
```
Why do we do this? The msg is invalid at this point.

jukkar replied:
We are receiving some garbage from outside of the device. I don't think we can trust the message if there is some invalid data in it but should discard it.

jukkar replied:
This was marked as resolved but IMHO it is not. If we receive garbage then we cannot continue with parsing as the data could be anything.

maass-hamburg replied:
This is a check, if the dhcp option 43 has encapsulated data. If the data of this option is less than 3, it can't be encapsulated, but it is still a valid option 43 send by the dhcpv4 server. Therefore we are treating it like a unknown option and skip it.

jukkar replied:
The error print in previous line implies that there is a error in parsing, we also return a `-EBADMSG` indicating that there was an error. Why would we trust the remaining data, it seems that rest of the data could be totally bogus in this case?

maass-hamburg replied:
@jukkar should we return `0` instant of `-EBADMSG` in these situations?

jukkar replied:
No, lets not change that. It is good that something is printed for the error case. I have been testing this with invalid data but have not been able to crash or get the parse to confuse itself. It does not mean that the parses works perfectly in all cases but lets deal with that when/if it happens.

m-byte commented:
@rlubos Yes, this is also something Fin and I discussed before he started working on this task. My reasoning here (and also why I wanted this to be a sub-option of option callbacks) is, that the vendor defining this option is the company or person writing the software for the embedded device. I.e. Zephyr would usually be the side controlled by the vendor. Due to this, the vendor can decide whether to use encapsulated options (enabling the corresponding sub-option in Kconfig) or whether to register a regular callback for option 43 instead.
One thing that is currently missing here (and on our todo list) is the transmission of option 60 (vendor class identifier) in the DHCP discover message. This is usually used by DHCP servers to decide what options to send out to a device. And that's kind of where the whole circle of vendor defined options closes.

As an aside: My main reason for wanting to parse encapsulated vendor specific options is that in this case only the data for a sub-option is copied at a time, rather than copying the whole option 43 once to then loop over it to actually copy the sub-options to where they are needed.

jukkar commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -667,6 +671,71 @@ static void dhcpv4_timeout(struct k_work *work)
+		length--;
+		if (length < len) {
+			NET_ERR("Vendor-specific option parsing, length too long");
+			net_pkt_skip(pkt, length);
```
If there is an error, we should discard the message and not try to do something like this. The packet could be formatted by a malicious actor so that it could lead to system compromise. After all, we are here parsing data that is coming from outside of the device and we should be very strict what we are accepting.

maass-hamburg replied:
With the code I added, I want to make sure 
1. that a dhcpv4 packet with vendor specific options, that is not encapsulated, will not break the program 
2.  that a non encapsulated packet will not result in a drop of the whole packet, which would result in not  getting a IPv4-Address, instant it will just ignore the vendor specific option and will go to the next option. 

To make extra sure we could add the cursor backup around the funktion. 
`net_pkt_cursor_backup(pkt, &backup);
				dhcpv4_parse_option_vendor(pkt, iface, msg_type, length);
				net_pkt_cursor_restore(pkt, &backup);`

jukkar replied:
If we can ignore safely the invalid vendor specific data, then that could be ok. I am a bit worried about doing `net_pkt_skip()` here, are we able to sync the parsing properly if there is a length error.

m-byte replied:
I started working on improving the tests to cover the most likely cases of parsing going out of sync (mostly invalid lengths in the encoded data). I'll add them later today

m-byte replied:
The tests now check for short packets in multiple cases, including the edge case of option 43 with a length of only 1. (I.e. only enough data for the option type and not even a length).
Fins code passes all of my tests.

m-byte committed:
tests: net: dhcpv4: tests for encapsulated vendor specific option callbacks

Add tests for the encapsulated vendor specific option callback API.
These test cases are focused on correct parsing of encapsulated
vendor specific options.
The sample DHCP messages have been extended accordingly.

Signed-off-by: Matthias Breithaupt <m.breithaupt@vogl-electronic.com>
Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>
Co-authored-by: Fin Maaß <f.mass@vogl-electronic.com>

m-byte commented on tests/net/dhcpv4/client/src/main.c:
```diff
@@ -137,18 +142,28 @@ static const unsigned char ack[382] = {
+0xfe, 0xfe, 0x2b, 0x01, 0xfe, 0x46, 0x04, 0xc6,
+0x33, 0x64, 0x10, 0xff
 };
 
```
That depends on what you mean by "invalid data in packets".
In this implementation, we have the following (total length refers to the option 43 length):
1. a properly encapsulated option 43, which contains the string "string" inside suboption 1
2. an encapsulated option with a total length of 15 with:
  a) properly encapsulated suboption 2 with a length of 1 and a value of 1 inside
  b) properly encapsulated suboption 3 with a length of 0 (i.e. no data)
  c) inproperly encapsulated suboption 254 with a length of 16, followed by 8 bytes of 254
3. an inproperly encapsulated option 43 with a length of 1 (i.e. too short to contain an encapsulated option) with 254 inside
4. option 70 with IP 198.51.100.16 (to make sure that processing is properly resumed)

Apart from checking the values of the correctly encapsulated options, there are callbacks registered for both option 254 and 43.254 that would cause an immediate failure of the test.

jukkar replied:
Ok, it was not clear from the tests what the data is. And one would need some rfc reading to decipher the packet data to understand what is being tested. Perhaps it would make sense to describe in the test the steps you have outlined above.

m-byte replied:
This should be addressed now. However, while doing so, I found what feels like a rather significant issue with DHCP tests:
Both arrays hold DHCP ACK messages. The one in `ack` contains no client IP and they differ in relay agent IP. I don't think this is intentional.

m-byte replied:
@maass-hamburg please check if there is any part of the code that overwrites some bytes to turn the offer into an actual offer. If not, use a wireshark dump of some actual interaction with a DHCP server to check what the expected behavior is (I'm not even sure the options are usually sent for both ack and offer). Since sending DHCP option 60 to the server will require slightly extended tests, it would make sense to fix the current test before starting work on option 60.

maass-hamburg replied:
all dhcp options in the dhcp packet data are now described by comments to make them easier to understand.

m-byte commented on tests/net/dhcpv4/client/src/main.c:
```diff
@@ -458,22 +561,64 @@ ZTEST(dhcpv4_tests, test_dhcp)
 
+#ifdef CONFIG_NET_DHCPV4_OPTION_CALLBACKS_VENDOR_SPECIFIC
+	net_dhcpv4_init_option_vendor_callback(&opt_vs_string_cb, vendor_specific_string_cb,
+					OPTION_VENDOR_STRING, buffer,
```
Seems that way because it is. Sorry

jukkar commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -667,6 +671,71 @@ static void dhcpv4_timeout(struct k_work *work)
+	uint8_t len;
+	uint8_t type;
+
+	if (length < 3) {
+		NET_ERR("Vendor-specific option parsing, length too short");
+		net_pkt_skip(pkt, length);
+		return -EBADMSG;
+	}
```
The more I look at this block, the more concerned I feel. Are we really able to recover from this if we receive a perfectly formatted message but where length is 2, like when a malicious actor sends a hand crafted message and sets the length 2 where the protocol expects some bigger value. Now we move the cursor two bytes forward and continue parsing the data from wrong position, or did I understand this totally wrong?

maass-hamburg replied:
the `length` here is the whole length of the option 43. Therefore if we move the cursor 2 bytes, we are going to the next option in the dhcpv4 packet, skipping the rest of the option 43. And there either the next dhcpv4 option (not a encapsulated option) should be or the end off the dhcpv4 packet with `0xff`

jukkar replied:
If the packet is malformed, then the parser will be out of sync and it is no point continuing parsing the message. I really think we should just discard it instead of trying to recover.

m-byte replied:
What we are trying to recover from here is the case where option 43 is misconfigured, which could e.g. happen if the DHCP server was set up for devices from a different manufacturer.
According to [RFC2132](https://datatracker.ietf.org/doc/html/rfc2132#section-8.4), "Clients which do not receive desired vendor-specific information SHOULD make an attempt to operate without it, although they may do so (and announce they are doing so) in a degraded mode.". So trying to ignore option 43 would be in line with this.
If the option packet itself is corrupted, as could be the case in an attack scenario, that is something we might want to look into regarding the DHCP client in general. However, for option 43, I added most of these special cases into the tests. If there is any case you can think of that isn't covered by the tests, let me know. (The one descibed above is covered)

jukkar replied:
Thanks for the clarifications and commenting the packet format in the test. I took your branch and experimented it a bit and it is super easy to get the dhcpv4 parser out of sync, but fortunately I did not see any real issues with it. Unknown options are just skipped, although it is probably possible to craft a message that allows valid options to be matched.

I think the dhcp option parser needs some looks to investigate if it is working properly, I can do that.

rlubos replied:
> What we are trying to recover from here is the case where option 43 is misconfigured, which could e.g. happen if the DHCP server was set up for devices from a different manufacturer.

Hmm, but all we do here is checking if the payload length is smaller than 3 bytes? How would that prevent from "misconfigured" Vendor option (I assume you mean here that the vendor did not use the encapsulated TLVs internally?). 
That's why I was opting for treating this vendor-information option as opaque at the DHCP client level, and let the application interpret it (probably in par with class-identifier option content).

m-byte replied:
Well, the easiest way of getting it out of sync are incorrect lengths. However, for my tests, I focused on the encapsulated option parser from this PR. And there, my main objective was making sure that for any incorrectly encoded option 43 the main parser would resume with the next option. For this I had to make the assumption that the lengths at the top level are correct. If they aren't it's not really related to this PR

jukkar replied:
I don't think we should do parsing differently in vendor options compared to normal dhcpv4 options. For example if there is length error for normal options we just drop the packet immediately. So I am proposing these additional changes here:
```
diff --git a/subsys/net/lib/dhcpv4/dhcpv4.c b/subsys/net/lib/dhcpv4/dhcpv4.c
index 4f9de7a3fbe..c75e90d76f6 100644
--- a/subsys/net/lib/dhcpv4/dhcpv4.c
+++ b/subsys/net/lib/dhcpv4/dhcpv4.c
@@ -683,7 +683,6 @@ static int dhcpv4_parse_option_vendor(struct net_pkt *pkt, struct net_if *iface,
 
        if (length < 3) {
                NET_ERR("Vendor-specific option parsing, length too short");
-               net_pkt_skip(pkt, length);
                return -EBADMSG;
        }
 
@@ -706,7 +705,6 @@ static int dhcpv4_parse_option_vendor(struct net_pkt *pkt, struct net_if *iface,
                length--;
                if (length < len) {
                        NET_ERR("Vendor-specific option parsing, length too long");
-                       net_pkt_skip(pkt, length);
                        return -EBADMSG;
                }
                net_pkt_cursor_backup(pkt, &backup);
@@ -882,9 +880,13 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
 #if defined(CONFIG_NET_DHCPV4_OPTION_CALLBACKS_VENDOR_SPECIFIC)
                case DHCPV4_OPTIONS_VENDOR_SPECIFIC: {
                        if (!sys_slist_is_empty(&option_vendor_callbacks)) {
+                               int ret;
+
                                NET_DBG("options_vendor_specific");
-                               if (dhcpv4_parse_option_vendor(pkt, iface, msg_type, length) ==
-                                   -ENOBUFS) {
+
+                               ret = dhcpv4_parse_option_vendor(pkt, iface, msg_type,
+                                                                length);
+                               if (ret == -ENOBUFS || ret == -EBADMSG) {
                                        return false;
                                }
                        } else {
```
According to the RFC, all the options should have a length field. I did not see mentions what should be done if length is wrong, but I would imagine that it is quite reasonable to drop the packet if there is some error that could compromise the integrity of the message.

maass-hamburg replied:
We only return `-EBADMSG` when the length inside of one vender specific option is wrong. I.e. we would otherwise go to far with the cursor (outside of the DHCP 43 option). But if we just have a non encapsulated vendor option like `0x2b, 0x04, 0x03, 0x0F, 0x10, 0x01`, which according to the RFC is still valid, and still have `CONFIG_NET_DHCPV4_OPTION_CALLBACKS_VENDOR_SPECIFIC` enabled we would drop the message, if we approve your change, because `0x0F` will be interpreted as a length of the first encapsulated option and it is longer than the field of the option 43, which is `0x04` long. In my opinion it is then wrong to drop the complete (and according to the RFC still valid)  packet. In that scenario we should instant  just ignore the content of the option 43 and go to the next option in the packet and continue. We can't know if the content of option 43 is encapsulated or not.

m-byte replied:
My computer is a bit weird today. Assume that all my comments in this thread belong one message before where they are shown. Sorry about that.

maass-hamburg replied:
RFC 2132 
>Clients which do not receive desired vendor-specific information
   SHOULD make an attempt to operate without it, although they may do so
   (and announce they are doing so) in a degraded mode.

According to this I would interpret this to say that we should still not drop the packet (which means not getting a ip lease) if we don't get the encapsulated vendor options, which we desire and instant get non encapsulated vendor option, which we don't desire. We should still get the ip lease. the same should happen, if the content of the dhcp option 43 is malformed (= not desired), as long as length of the option 43 is right .

rlubos replied:
Well, to be honest - I'm not convinced. It looks to me that all the logic around returning `EBADMSG` is just trying to circumvent the fact, that Vendor information option does not have a consistent format **and should be left** to the vendor-specific code to interpret (so not trying do to so in the generic library), as the RFC says:
> The information is an opaque object of n octets, **presumably interpreted by vendor-specific code on the clients
   and servers**.

Parsing the content in a generic part of the library, **assuming** that the option may have sub-TLVs (or may not), is IMO just a source of confusion in this thread.

jukkar replied:
> Parsing the content in a generic part of the library, **assuming** that the option may have sub-TLVs (or may not), is IMO just a source of confusion in this thread.

Indeed, this seems to be the case here. Could we leave the vendor option parsing to application callback?

maass-hamburg replied:
If the vendor (in the RFC sense) decides to make use of encapsulated vendor specific options, they activate the corresponding Kconfig option. Then, they can use the provided functions that decode their encapsulated options in kernel space.
Like other functions provided by Zephyr, it is up to the vendor to activate them. For example, Zephyr provides lots of drivers for sensors, which are also not used by everyone - they also have to be activated by the vendor.
If a vendor doesn't want to use encapsulated vendor options, nothing prevents them from doing so. Without the Kconfig set, none of the code from this PR is compiled in.
 
With this PR, we are providing an easy and memory efficient way of using encapsulated vendor options that can be easily integrated into products. A main advantage of this is, that it makes it less likely for vendors to abuse "top level" options for their purposes.
Encapsulated vendor specific options are already being used in lots of places, SIP phones and thin clients being great examples. If someone were to implement one using Zephyr, this would be a great foundation for the DHCP side of things.
Due to their use by various products, they are already supported by various DHCP servers (e.g. MS DHCP, ISC Kea, dnsmasq) and wireshark has decoders for them built in. In the case of wireshark, option 60 is used to differentiate them. However, I would attribute this to the fact that wireshark has to be able to dissect encapsulated options for lots of vendors. On the other hand, it would certainly be possible to add a check for an expected value in option 60 before running this decoder. (obviously, that would depend on sending option 60 to the server first)

jukkar replied:
> If a vendor doesn't want to use encapsulated vendor options, nothing prevents them from doing so. Without the Kconfig set, none of the code from this PR is compiled in.

Sure, this is business as usual and nothing new here. Both Robert and I are concerned about the implementation as it seems that the actual vendor extension parsing would better be done in the code that actually knows about the exact format of the data.

So would it make sense that instead of this where we try to parse semi opaque data,
```
	while (!net_pkt_read_u8(pkt, &type)) {
		if (type == DHCPV4_OPTIONS_END) {
			NET_DBG("Vendor-specific options_end");
			return 0;
		}
		length--;

		if (length <= 0) {
			NET_ERR("Vendor-specific option parsing, malformed option");
			return -EBADMSG;
		}

		if (net_pkt_read_u8(pkt, &len)) {
			NET_ERR("Vendor-specific option parsing, bad length");
			return -ENOBUFS;
		}
		length--;
		if (length < len) {
			NET_ERR("Vendor-specific option parsing, length too long");
			return -EBADMSG;
		}
		net_pkt_cursor_backup(pkt, &backup);

		SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&option_vendor_callbacks, cb, tmp, node) {
			if (cb->option == type) {
				NET_ASSERT(cb->handler, "No callback handler!");

				if (net_pkt_read(pkt, cb->data, MIN(cb->max_length, len))) {
					NET_DBG("option vendor callback, read err");
					return -ENOBUFS;
				}

				cb->handler(cb, len, *msg_type, iface);
				net_pkt_cursor_restore(pkt, &backup);
			}
		}
		net_pkt_skip(pkt, len);
		length = length - len;
		if (length <= 0) {
			NET_DBG("Vendor-specific options_end (no code 255)");
			return 0;
		}
	}
```
we leave the parsing of the vendor specific data to the callback. We could do something like this (simplified example), so giving the whole vendor option directly to the callback to process it?
That way we also do not need to care what the format of the data is.
```
               bytes_copied = net_buf_linearize(user_supplied_buffer, user_supplied_buffer_len,
					pkt->frags, data_offset, length);

		SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&option_vendor_callbacks, cb, tmp, node) {
			if (cb->option == type) {
				NET_ASSERT(cb->handler, "No callback handler!");

				cb->handler(cb, user_supplied_buffer, bytes_copied, iface);
			}
		}
```
WDYT?

maass-hamburg replied:
If I understand your proposal right, you just want to give back one callback for the whole content of the dhcp option 43. If that's what you mean, I don't see the benefit then instant just using the regular option call and register a callback for option 43. 
But that would still be possible to do with my commit. If the vendor doesn't want to use the encapsulated vendor specific option like my commit provides, he can still use the regular option callbacks and register a callback for the option 43 and handle the parsing in application code. But with my commit we enable the vendor to use the vendor specific options as they are recommended in the RFC, but hey don't have to.

jukkar replied:
> But that would still be possible to do with my commit. If the vendor doesn't want to use the encapsulated vendor specific option like my commit provides, he can still use the regular option callbacks and register a callback for the option 43 and handle the parsing in application code. But with my commit we enable the vendor to use the vendor specific options as they are recommended in the RFC, but hey don't have to.

Indeed that is possible, so my proposal is not really needed. 

But imho my original concern is still valid, if there is a protocol error we should abort the parsing instead of trying to recover.

rlubos replied:
> malformed encapsulated vendor specific option is still a valid vendor specific option, as long as the length of the whole option 43 is right

But what if this "malformed" option is indeed a valid vendor option from a different vendor that does not use encapsulated option encoding, but the first byte of their opaque data matches the suboption type we've registered a callback for? From the parser side, all could seem right, but we actually report rubbish to the application. Or even if some different vendor still uses encapsulated encoding, they could redefine the option codes as they wish, as it's allowed by RFC 1533:
> Codes other than 0 or 255 MAY be redefined by the vendor within the encapsulated vendor-specific extensions field.

That's why (I'm repeating myself), in my opinion any attempt to parse Vendor Specific option content (including assuming it uses encapsulated option encoding) should only take place after identifying whether we deal with a particualr vendor, as RFC 1533 says:
> The definition of this information is vendor specific. The vendor is indicated in the class-identifier option.

Otherwise, after enabling this feature, we're practically binding the generic implementation here to a single vendor. I may still work with other vendors, but could produce unexpected outcomes as described in the first paragraph.

m-byte replied:
On more thought: assuming the DHCP server also sends back the clas identifier (don't know off the top of my head), what if there were only a single option 43 callback and rather than registering individual callbacks, there would just be buffers for the output of each suboption and a single callback, that callback could be only fired once an expected class identifier and a decodable option 43 have been received. Would that approach suit you better?

jukkar commented on include/zephyr/net/dhcpv4.h:
```diff
@@ -163,6 +163,48 @@ int net_dhcpv4_remove_option_callback(struct net_dhcpv4_option_callback *cb);
 
```
BTW, just noticed this. Could you use lower case letters in commit subject, so `net: dhcpv4: ....`, same for the testing commit, which should be called `tests: net: dhcpv4: ....`

jukkar commented on tests/net/dhcpv4/client/src/main.c:
```diff
@@ -36,8 +36,8 @@ LOG_MODULE_REGISTER(net_test, CONFIG_NET_DHCPV4_LOG_LEVEL);
 #define NET_LOG_ENABLED 1
```
One minor nit, please change the commit subject of the test patch to `tests: net: dhcpv4: ....` so that it is easier to find when grepping the git log.

maass-hamburg committed:
tests: net: dhcpv4: add comments to sample packets

To make the content of the sample packets more more
understandable, comments describing it are added.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

rlubos reviewed (dismissed):
Given recent comments, I think we're good with the current approach.

maass-hamburg commented:
rebase to resolve merge conflicts
